sklearn/metrics/cluster/tests/test_unsupervised.py

```diff
@@ -228,3 +228,26 @@ def test_davies_bouldin_score():
     X = ([[0, 0], [2, 2], [3, 3], [5, 5]])
     labels = [0, 0, 1, 2]
     pytest.approx(davies_bouldin_score(X, labels), (5. / 4) / 3)
+
+
+def pairwise_distances_nonzero_diag(X, Y=None, metric="euclidean", n_jobs=None, **kwds):
+    # Intercept pairwise distances and corrupt the diagonal entries
+    dists = pairwise_distances(X, Y, metric, n_jobs, **kwds)
+
+    if len(np.squeeze(dists).shape) > 1:
+        np.fill_diagonal(dists, 1)
+    return dists
+
+
+def test_davies_bouldin_nonzero_diag():
+    # Test that davies-boulding is invariant to non-zero diagonals on distance matrices
+    dataset = datasets.load_iris()
+    X = dataset.data
+    y = dataset.target
+
+    original_score = davies_bouldin_score(X, y)
+
+    nonzero_diag_score = davies_bouldin_score(
+        X, y, pairwise_distances_func=pairwise_distances_nonzero_diag)
+
+    assert_equal(original_score, nonzero_diag_score)
```

Comment:
Are you sure we want strict equality, not floating point approximate equality? If strict equality, we should use an assert statement, not `assert_equal` since we have adopted pytest

Reply:
The patch modifies the diagonal entries of the distance matrix so that they are exactly 0. So the output should (I think) be exactly the same therefore we should test for strict equality. Any deviation away from strict equality would indicate that the non-zero entries have perturbed the score somehow.

Reply:
You might be right. You just always need to be careful in the use of exact equality with floats: different processors use different orders for summation, etc.